pyrobosim_ros/test/test_ros_interface.py

```diff
@@ -0,0 +1,248 @@
+        rclpy.init()
+
+        # Load world from file.
+        world_file_path = os.path.join(get_data_folder(), "test_world_multirobot.yaml")
+        world = WorldYamlLoader().from_yaml(world_file_path)
+
+        # Create ROS interface.
+        TestRosInterface.ros_interface = WorldROSWrapper(
+            state_pub_rate=0.1, dynamics_rate=0.01
+        )
+        TestRosInterface.ros_interface.set_world(world)
+        TestRosInterface.ros_interface.start(auto_spin=False)
+
+        TestRosInterface.node = Node("test_ros_interface")
+
+        TestRosInterface.executor = MultiThreadedExecutor()
+        TestRosInterface.executor.add_node(TestRosInterface.ros_interface)
+        TestRosInterface.executor.add_node(TestRosInterface.node)
+        time.sleep(1.0)
+
+    @staticmethod
+    @pytest.mark.dependency(
+        name="test_robot_pub_sub", depends=["test_start_ros_interface"]
+    )
+    def test_robot_pub_sub():
+        """Test that we can publish and subscribe to the robots."""
+
+        data = {"latest_state": None}
+
+        def state_cb(msg):
+            data["latest_state"] = msg
+
+        state_sub = TestRosInterface.node.create_subscription(
+            RobotState, "robot0/robot_state", state_cb, 10
+        )
+        vel_pub = TestRosInterface.node.create_publisher(Twist, "robot0/cmd_vel", 10)
+
+        start_time = time.time()
+        while not data["latest_state"]:
+            TestRosInterface.executor.spin_once(timeout_sec=0.1)
+            if time.time() - start_time > 2.0:
+                break
+
+        assert data["latest_state"] is not None
+        init_state = data["latest_state"]
+
+        vel_msg = Twist()
+        vel_msg.linear.x = 0.1
+        vel_msg.angular.z = 0.5
+        for _ in range(25):
+            vel_pub.publish(vel_msg)
+            TestRosInterface.executor.spin_once(timeout_sec=0.1)
+            time.sleep(0.01)  # Give the publisher a break
+
+        assert data["latest_state"] != init_state
+
+        TestRosInterface.node.destroy_subscription(state_sub)
+        TestRosInterface.node.destroy_publisher(vel_pub)
+
+    @staticmethod
+    @pytest.mark.dependency(name="test_get_world_state", depends=["test_robot_pub_sub"])
+    def test_get_world_state():
+        """Test that we can retrieve the world state via service call."""
+
+        client = TestRosInterface.node.create_client(
+            RequestWorldState, "request_world_state"
+        )
+
+        future = client.call_async(RequestWorldState.Request())
+        start_time = time.time()
+        while not future.done():
+            TestRosInterface.executor.spin_once(timeout_sec=0.1)
+            if time.time() - start_time > 2.0:
+                break
+
+        assert future.done()
+        result = future.result()
+        assert len(result.state.robots) == 3
+        assert len(result.state.locations) == 4
+        assert len(result.state.objects) == 8
+
+        TestRosInterface.node.destroy_client(client)
```

Comment:
What happens if this prematurely exists? Will the test clean up everything, including the node, before moving on?

Reply:
Yeah, the node will definitely get destructed if the program exits!